File: fsharp/binary-search/BinarySearch.fs
module BinarySearch

let find input value =
    
    let rec loop (array: int array) value low high =
        if low > high then None
        else
            let pivot = (low + high) / 2
            let item = array[pivot]
            
            match item with
            | elem when value = elem ->
                Some pivot
            | elem when value > elem ->
                loop array value (pivot + 1) high
            | elem when value < elem ->
                loop array value low (pivot - 1)
            | _ ->
                None
               
    
    if (input: int array).Length = 0 then None
    else
       loop input value 0 (input.Length - 1)

File: go/binary-search/binary_search.go
package binarysearch

func SearchInts(list []int, key int) int {
	var mid = 0
	for lo, hi := 0, len(list)-1; lo <= hi; {
		mid = (lo + hi) / 2
		if list[mid] < key {
			lo = mid + 1
		} else if list[mid] > key {
			hi = mid - 1
		} else {
			return mid
		}
	}

	return -1
}
